[1 - 2단계 방탈출 예약 대기] 메이(김시원) 미션 제출합니다.

seaniiio opened the pull request:
안녕하세요 피케이!

리뷰 잘 부탁드립니다 😀

[변경사항 링크입니다!](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/files/c00678105a8aaef24b6a1ae02205eddb94c3eea2..5aec888ee5f03eeec178fd925901f53c048d738e)

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [프롤로그 링크입니다!](https://prolog.techcourse.co.kr/studylogs/4382)
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

- JPA를 처음 사용해봤는데, 전환이 제대로 되었는지 궁금합니다!

- Mockito를 이용하여 서비스 계층을 테스트했습니다. 기존에는 FakeRepository(FakeDao)를 구현해서 테스트에 사용했는데, join 기능을 직접 구현하는데 어려움이 있었고, Service의 비즈니스 로직을 테스트하기 위한 최소한의 stubing을 해주기 위해 Mockito를 적용했습니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

changuii committed:
init: 프로젝트 가져오기

changuii committed:
refactor: RDB 기반의 repository를 JPA 기반으로 전환

기존 JdbcTemplate을 활용하던 repository 구현체를 제거하고, JPA repository로 리팩토링하여 코드 단순성과 유지보수성을 개선하였습니다. 이에 따라 관련 도메인 엔티티 클래스에 JPA 매핑 어노테이션(@Entity 등)을 추가하였으며, 연관 관계를 매핑하여 관리하도록 수정하였습니다. 추가적으로 사용되지 않는 schema.sql 제거 및 예약 관련 유효성 검사 로직을 간소화하였습니다.

changuii committed:
refactor: @AllArgsConstructor를 활용하여 생성자 제거 및 불필요한 @Autowired 어노테이션 삭제

- Lombok의 @AllArgsConstructor를 추가하여 생성자를 자동 생성하도록 변경
- @Autowired 어노테이션 삭제 및 관련 수동 생성자 제거
- 클래스 필드의 final 제거, 가변성을 고려한 수정
- 코드 가독성과 유지보수성을 향상시킴

changuii committed:
test: Mock 기반으로 Fake Repository 제거 및 테스트 리팩토링

changuii committed:
feat: reservation-mine 페이지를 응답하라

changuii committed:
test: reservation-mine 페이지를 응답하라

changuii committed:
feat: 예약 상태를 나타내는 ReservationStatus 추가 및 멤버별 예약 조회 기능 구현

- `ReservationStatus` enum 추가 및 예약 상태 관리
- `MemberReservationResponse` 클래스 추가로 예약 조회 DTO 정의
- 멤버별 예약 조회 엔드포인트(`/members/reservations`) 및 서비스 로직 추가
- `Member` 엔티티에 `reservations` 연관관계 설정
- SQL 스키마에 `reservation_status` 필드 추가 및 초기 데이터 수정

seaniiio committed:
refactor: 엔티티 id 필드의 final 제거

pkeugine commented on src/main/java/roomescape/theme/ThemeRepository.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.data.jpa.repository.Query;
+
+public interface ThemeRepository extends JpaRepository<Theme, Long> {
+    @Query(nativeQuery = true,
+            value = """
+                            SELECT t.id, t.name, t.description, t.thumbnail
+                            FROM theme t INNER JOIN reservation r
+                                ON t.id=r.theme_id
+                            WHERE r.date BETWEEN ?1 AND ?2
+                            GROUP BY t.id, t.name, t.description, t.thumbnail
+                            ORDER BY COUNT(*) DESC
+                            LIMIT ?3
+                    """)
```
```suggestion
    @Query(nativeQuery = true,
            value = """
                            SELECT t.id, t.name, t.description, t.thumbnail
                            FROM theme t INNER JOIN reservation r
                                ON t.id=r.theme_id
                            WHERE r.date BETWEEN :from AND :to
                            GROUP BY t.id, t.name, t.description, t.thumbnail
                            ORDER BY COUNT(*) DESC
                            LIMIT :size
                    """)
```
가독성을 위해 쿼리 내부에서 파라미터 값을 이름 기반으로 받아오는건 어떨까요?

seaniiio replied:
이름 기반 바인딩으로 수정하겠습니다!

[refactor: sql 쿼리 이름 기반 바인딩으로 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/a689eb74c191c69f89f788b143d9e78f30d2fbfc)

pkeugine commented on build.gradle:
```diff
@@ -21,10 +21,15 @@ dependencies {
+    implementation 'io.jsonwebtoken:jjwt-impl:0.11.5'
+    implementation 'io.jsonwebtoken:jjwt-jackson:0.11.5'
+
+    compileOnly 'org.projectlombok:lombok:1.18.38'
+    annotationProcessor 'org.projectlombok:lombok:1.18.38'
```
롬복을 사용하셨군요.
그런데 롬복 의존성을 추가한 커밋이 [refactor: ThemeRepository를 JpaRepository...(후략)](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/9c91e5142f27087ad13df28e4e817fce916fa8b7) 입니다.

이왕이면 [롬복 로직이 포함되는 커밋](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/645a26313466b06da9d3a3cf53595cd5ac16276a) 또는 의존성 추가만 하는 별도의 커밋으로 구분하면 어땠을까 합니다.

이제 프로젝트 고도화를 해보고 있고, 깃도 많이 사용해보셨으니 커밋 관리를 잘 해보는 경험을 해보는 것도 좋을 것 같네요.
rebase 를 활용하여 커밋을 분리할 수 있는데, 이를 수정해보는 것을 추천드립니다.
(너무 시간이 오래걸린다면 해당 리뷰는 패스하셔도 괜찮습니다.)

seaniiio replied:
rebase 이용해서 의존성 추가 커밋을 분리했습니다!

앞으로의 미션에서는 커밋 관리에 더 신경써보도록 하겠습니다.

pkeugine commented on src/main/java/roomescape/auth/AuthController.java:
```diff
@@ -0,0 +1,43 @@
+import roomescape.auth.dto.LoginResponse;
+
+@RestController
+@AllArgsConstructor
```
롬복을 사용하셔서 궁금해지네요.
`@AllArgsConstructor` 와 같은 롬복 기능을 사용해서 얻은 이점이 있었나요?

더 나아가, 롬복을 사용할 때 주의해야 할 점과 단점도 있습니다.
메이가 생각하시는 이러한 단점이 무엇인지 궁금합니다.

seaniiio replied:
## 롬복을 사용하면서 느낀 장점

- 엔티티마다 반복적으로 작성해줘야 했던 보일러플레이트 코드를 생략할 수 있다는 것이 큰 장점으로 느껴졌습니다. JPA를 사용하기에 기본 생성자가 꼭 필요했고, 용이한 테스트를 위해 전체 필드 생성자도 필요했습니다. 그래서 모든 엔티티에 `@NoArgsConstructor`, `@AllArgsConstructor`를 사용해도 되겠다고 생각했어요.
- 생성자, getter, equals를 애노테이션으로 대체하니 클래스 내의 코드 양이 줄어들었습니다. 지금은 엔티티에 별다른 로직이 없지만, 만약 엔티티에 비즈니스 로직이 추가된다면 중요한 로직에 집중하기 좋을 것 같아요.

## 롬복을 사용하면서 느낀 단점 / 주의할 점

사실 사용하는 과정에서는 단점을 느끼지 못했는데.. 열심히 고민해봤습니다. 

- 정말 꼭 필요한 기능이 아닐 수 있는데, 무의식적으로 애노테이션으로 선언해버리는 것에 대해 주의해야겠다는 생각이 들었습니다. 사실 AllArgsConstructor는 JPA로 전환하는 과정에서 테스트 이외에는 사용하지 않게 되었는데, 롬복을 이용하다보니 무의식적으로 만들었습니다. (테스트에만 사용하는 생성자이긴 하지만, 테스트 용이성을 위해 남겨두는 것이 좋겠다고 판단해서 남겨뒀습니다!)

```java
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Member {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @EqualsAndHashCode.Include
    private Long id;
```

- 가독성을 위해 롬복을 사용했는데, 오히려 읽기 불편해질 수 있을 것 같습니다. 예를 들어 저는 이번에 필드에 `@EqualsAndHashCode.Include`를 붙여주는 방식으로 `equals()`를 구현했습니다. 그런데 다른 크루들이 이 코드를 보면서 “이건 뭐지..?” 라고 했던 것이 기억에 남아요. 롬복은 공식 자바 문법이 아닌 라이브러리이기 때문에, 협업하는 사람들이 읽기 편한지도 충분히 고려하면서 사용해야겠다는 생각이 들었습니다.
    
    +) `@EqualsAndHashCode(of = {})` 방식을 사용하지 않은 이유는, 필드 이름을 바꿀 때 of 속성도 수정해야 하는 것을 방지하기 위함입니다!
    
- 호기심 많은 페어 덕분에 롬복 라이브러리를 까보면서 어떤 기능이 있는지 직접 확인하는 시간을 가졌습니다. 제가 기존에 어렴풋이 알고 있던 것보다 훨씬 많은 기능을 제공하고 있었어요. 이를 잘 활용하면 코드가 깔끔해지겠지만, 익숙하지 않은 입장에서 이해가 어려울 수 있을 것 같고, 이 부분을 주의해야 할 것 같습니다.

## 찾아본 단점

제가 위에 언급한 단점은 아주 기본적인 getter, 생성자를 만드는 정도만 잘 생각하며 활용하면 충분히 커버 가능할 것 같습니다. 추가적으로 단점 및 주의할 점을 찾아본 결과 아래와 같았습니다.

- `@Data`, `@ToString`, `@EqualsAndHashCode` 사용 시 양방향 관계에서 순환 참조 발생 가능
- 롬복이 만들어주는 생성자는 필드 순서에 의존 → 필드 순서 바뀌면 예기치 못한 상황 발생 가능
- getter, setter, 생성자 등에 커스텀 로직 삽입할 수 없음
- 컴파일 시간 증가 가능성
- IDE 플러그인에 크게 의존적임

…

페어가 제안해줘서 적용했고 너무 편하다고 느꼈는데, 조심해야 할 부분이 생각보다 많다는 것을 알게 되었습니다. 그럼에도 보일러플레이트 코드를 제거해준다는 장점이 크다고 느껴서, 롬복 사용을 유지하되, 주의하며 사용해보도록 하겠습니다!

pkeugine replied:
잘 알아봐주셨네요.
메이가 정리해준 내용 모두 공감되고,
조금 더 적어보자면 제대로 잘 알고 사용하지 않으면 디버깅 할 때 불편해질 수 있다는 점이 있습니다.
하지만 충분히 학습하고 알아보며 사용하시는 듯 하여 큰 걱정은 없을 것 같네요.
이제 이론적인 내용은 알았으니 사용해보며 직접 체험하면 될 것 같습니다. 👍

pkeugine commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1,8 @@
+spring.datasource.url=jdbc:h2:mem:database;
+spring.jpa.show-sql=true
+spring.jpa.properties.hibernate.format_sql=true
+spring.jpa.ddl-auto=create-drop
```
`spring.jpa.ddl-auto` 설정을 create-drop 으로 해주신 이유가 궁금합니다.

seaniiio replied:
사실 미션의 JPA 설정 가이드를 그대로 긁어온거라, 딱히 이유가 없습니다. 그런데 create 설정만 해줘도 테이블 생성 전에 drop을 다 해주는데, create-drop의 필요성이 뭔지 의문이 들었고 이에 대해 생각해봤어요.

엔티티 설계가 바뀌어서 `C 엔티티`가 사라지는 경우, 애플리케이션 재시작 시 `C 테이블`을 drop하지 않아서 메모리 낭비가 있을 수 있겠다는 생각이 들었습니다. *그런데 어차피 H2쓰는데 애플리케이션 꺼지면 db 자동으로 사라지는거 아닌가?* 라고 생각했는데, h2가 파일 모드면 애플리케이션이 종료되어도 테이블이 그대로 남아있다는 것을 알게 됐어요. h2 db를 사용할 때는 이러한 경우를 생각해서 create-drop을 사용하는게 아닌가 싶었습니다.

메모리 모드를 사용하고 있는 지금은 create나 create-drop이나 뭘 사용해도 상관 없을 것 같습니다.

pkeugine replied:
이런 프로젝트성 애플리케이션이나 실전에서의 극초반 경우가 아니라면 `create-drop` 또는 `create` 는 거의 사용하지 않습니다.
하지만 반대로 말하면 그 두 경우에는 굉장히 유용한 옵션이라고 생각해요.
좋은 답변입니다 :)

pkeugine commented on src/main/java/roomescape/theme/ThemeRepository.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.data.jpa.repository.Query;
+
+public interface ThemeRepository extends JpaRepository<Theme, Long> {
+    @Query(nativeQuery = true,
```
현재 `Member`와 `Theme`의 연관 관계가 단방향으로 설정되어 있기 때문에, `JPQL`로 `Theme`을 기준으로 조회하면서 `Reservation`까지 함께 가져오는 것이 어려워 `NativeQuery`를 사용하신 것으로 추측됩니다.
제가 추측한 의도가 아니라면 말씀해주세요 :)

현재 프로젝트에 양방향 매핑과 단방향 매핑을 함께 사용해주시고 계신데요,
누군가는 "하나의 방식으로 통일" 하자고 말할 수도 있습니다.
그럴 때 메이는 어떻게 말할지 궁금합니다.

seaniiio replied:
테이블을 기반으로 작성하는 native query가 더 익숙하기 때문에 native query로 작성했었습니다. 그런데 **repository 인터페이스**에 작성하는 쿼리는 db 테이블보다 도메인 엔티티에 의존하는 것이 좋겠다는 생각이 들어서, JPQL로 수정하겠습니다!

[refactor(ThemeRepository): @Query에서 native query 대신 JPQL 사용하도록 수정](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/d8328dfcfa96781be1b25628d14406e652508566)

### 양방향 연관관계에 대한 생각

- 위에서 롬복 사용 시 주의점에 대해 알아보다가, `@ToString`, `@EqualsAndHashCode`를 사용할 때, 양방향 연관관계가 존재할 경우 순환참조를 주의해야 한다는 것을 알게 되었습니다. 또한 일관성을 맞추기 위해 편의 메서드를 작성해야 한다는 것도 단점이라 생각합니다.
- 다만 양방향 연관관계를 설정하면 분명 편하다는 장점이 있었습니다. `Member`별 `Reservation`을 불러와야 할 때 `MemberRepository`에서 바로 불러올 수 있다는 것이 장점이라 생각했어요. 만약 양방향 연관관계를 없앤다면 `MemberService`에 `ReservationRepository`에 대한 의존성도 추가해야 합니다.
- 양방향 연관관계는 순환참조에 대한 위험성, 편의 메서드 작성이 필수적이라는 점에서 단점이 더 크다고 느꼈습니다. 그리고 단방향, 양방향 중 한 가지 방식으로 통일하는 것이 더 일관적일 것 같아서, 단방향으로 통일하도록 하겠습니다.

이 부분에 대해서는 더 고민해보겠습니다! 현재는 단방향을 사용하는 것이 더 안전하기에 선택했는데, 추후 생각이 바뀌어 다시 양방향 연관관계로 수정하게 될 수 있을 것 같습니다.

[refactor: Member에서 Reservation으로의 일대다 단방향 연관관계 제거](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/70d4ec3d1e66c2cb0036b6bd36d6786731150073)

pkeugine replied:
네, 직접 시도해보며 체감하는게 제일 좋은 학습법이라고 생각합니다.
또한 꼭 통일하자는 리뷰는 아니었습니다. 두 방식을 모두 가져가는 경우도 많으니까요.
그래서 메이라면 어떻게 말할지~ 느낌으로 리뷰를 남겼던 것이고요.
잘 학습해주고 계십니다.

pkeugine commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @EqualsAndHashCode.Include
+    private Long id;
```
id 만을 사용하여 `@EqualsAndHashCode` 를 만들어주신 이유가 궁금합니다.
틀렸다는 의미는 아니고 생각해보자는 취지에서 남기는 리뷰입니다.

seaniiio replied:
`equals()`는 객체가 의미적으로 동등한 존재인지 확인하는 메서드인데, Entity에서 id의 존재 자체가 애초에 동등성을 확인하기 위함이라고 생각했습니다.

Entity는 id가 존재하는 도메인 객체이고, id가 필요한 이유는 객체를 식별하기 위함입니다. 레벨1에서는 애플리케이션이 돌아가는 동안만 동작하면 됐기 때문에 id가 필요하지 않았습니다. 그런데 서버를 재시작해도 데이터가 남아있어야 하는 상황을 고려한다면 객체를 주소값으로 식별할수 없기 때문에, 객체를 식별할 수 있는 id의 존재가 필요해졌습니다. 그리고 `equals()`는 객체가 의미적으로 동등한(equal) 존재인지 확인하는 역할을 하기에, 식별의 이유로 만든 Id를 이용하는 것이 맞다고 생각합니다.

pkeugine commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -0,0 +1,54 @@
+    @Enumerated(EnumType.STRING)
+    private MemberRole role;
+
+    @OneToMany(mappedBy = "member")
+    private final Set<Reservation> reservations;
+
+    public Member() {
+        reservations = new HashSet<>();
+    }
```
```suggestion
    @OneToMany(mappedBy = "member")
    private final Set<Reservation> reservations = new HashSet<>();

    protected Member() {
    }
```

엔티티의 생성 방식을 저라면 이 방식으로 리팩터링 해볼 것 같습니다.
어떤 차이가 있을지 생각해보고, 만약 메이의 방식을 유지하고자 한다면, 그 이유를 알려주세요.

이왕이면 해당 코멘트는 ai 에게 어떤 차이가 있는지 바로 알아보기보단
여태 미션들에서 배운 내용들을 토대로 생각해보는 재미를 느껴보셨으면 합니다.

seaniiio replied:
현재 `Member`에서 `Reservation`의 일대다 연관관계를 삭제하여 reservations 필드는 사라졌지만, 남겨주신 리뷰의 의도에 대해 생각해봤습니다.

JPA에서는 기본 생성자를 사용해서 객체 껍데기를 만든 후, db에서 가져온 값을 리플렉션으로 꽂아넣기 때문에, 기본 생성자 body에 아무것도 적어주지 않는 것이 혼란이 없을 것 같습니다. 혹시 피케이가 필드 초기화로 수정하신 의도가 이 이유가 맞을까요? 아니라면 지적해주시면 감사하겠습니다 😀

또한 기본 생성자는 JPA에서 사용할 수 있도록 열어둔건데, JPA에서는 `protected`이어도 접근할 수 있어서, 기본 생성자를 `protected`으로 선언하는게 좋을 것 같습니다.(이 부분은 스스로 생각해낸 것은 아니고, 얼마 전에 gpt와 대화를 나누다가 알게 됐습니다.)

```java
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Member {
```

[refactor: @NoArgsConstructor 접근 제한을 protected로 설정](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/5489872d15310c532c8573937381c01a335a2056)

pkeugine replied:
네, 잘 적어주셨고
no argument constructor 의 경우 실제 로직에서는 사용하지 않을테니
실수나 혼란을 방지하기 위해 `protected` 처리 하는게 좋다고 생각했습니다.

pkeugine commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -0,0 +1,54 @@
+        reservations = new HashSet<>();
+    }
+
+    public Member(final Long id, final String email, final String password, final String name, final MemberRole role) {
+        this.id = id;
+        this.email = email;
+        this.password = password;
+        this.name = name;
+        this.role = role;
+        this.reservations = new HashSet<>();
+    }
```
테스트에서만 사용하는 생성자를 없애보면 어떨까요?

seaniiio replied:
테스트에서 id가 없는 엔티티를 사용하면 NPE가 발생하는 경우가 존재해서, id까지 세팅해주는 생성자를 엔티티에 만들었습니다. 생성자를 테스트에만 사용한다 해도 테스트에 꼭 필요하기 때문에 열어두는게 맞다 생각했어요.(위의 리뷰 답글에도 이렇게 적었었습니다) 그런데 확실히 엔티티에 id를 주입해주는 것은 JPA의 역할이고, JPA는 기본 생성자 + Reflection으로 id를 주입해주는데, id 포함 생성자를 열어두는 것은 위험성이 있는 것 같아요.

id가 있어야 하는 이유는, 엔티티를 dto로 파싱하는 과정에서 NPE가 발생할 수 있기 때문입니다. 그래서 이 부분에 대해 어떻게 해결하는게 좋을 지 고민해봤어요.

### 1. reflection으로 id 세팅해주기
jpa처럼 reflection으로 Id값을 세팅해주는 코드를 테스트에 작성하는 방법이 있을 것 같습니다. 테스트에서만 사용하기에 문제가 되지 않을 것 같습니다.
### 2. entity의 id가 없어도 dto로 파싱 가능하도록 하기
dto 파싱 코드를 수정해야 합니다. 이또한 테스트를 위한 코드가 될 수 있을 것 같아요.

일단 1번 방법을 이용하도록 하고, 모든 엔티티에서 `@AllArgsConstructor`를 제거했습니다!

[refactor(Reservation): @AllArgsConstructor 제거](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/f88d4e28bd1a77845094ecf15f7e83401fcaa634)
[refactor(ReservationTime): @AllArgsConstructor 제거](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/0a19280a7a7cd638ed9e67a6214560780d195875)
[refactor(Theme): @AllArgsConstructor 제거](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/9fd28377004822e207efe074272b33993662b379)
[refactor(Member): @AllArgsConstructor 제거](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/9b4064ea02ab9e9294926fd7ceae7b5a01b0c4f4)

pkeugine commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -0,0 +1,54 @@
+    @OneToMany(mappedBy = "member")
+    private final Set<Reservation> reservations;
+
+    public Member() {
+        reservations = new HashSet<>();
+    }
```
lombok 에는 사용해주신 `@AllArgsConstructor` 외에도 `@NoArgsConstructor` 라는 것이 있습니다.
오히려 JPA 의 작동방식 때문에 생겨나는 이런 생성자야말로 롬복으로 만들어 볼 수 있지 않을까 싶네요.

seaniiio replied:
모든 엔티티에 `@NoArgsConstructor`를 선언해주도록 하겠습니다!

pkeugine commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -0,0 +1,48 @@
+public class Reservation {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
`GenerationType` 에는 어떤 것들이 있나요?
`IDENTITY` 를 선택해주신 이유도 궁금합니다.

seaniiio replied:
- IDENTITY → db가 생성한 id를 자동으로 세팅합니다. 
- AUTO → 사용하는 db 종류에 따라 알아서 설정합니다.
- TABLE → id 테이블을 만들어 관리합니다
- SEQUENCE → db 시퀀스 값으로 id를 먼저 파악한 후, Id가 포함된 객체를 생성합니다.

AUTO를 이용하면 다양한 db에 대해 유연한 설정이 가능하다 생각해서, 처음에는 AUTO를 사용했습니다. 그런데 db 및 데이터 초기화 과정에서 계속 에러가 떴고, IDENTITY로 수정하니 정상 동작했습니다. 이후 그 이유를 찾아보니, H2는 시퀀스를 제공하기 때문에 AUTO 사용 시 SEQUENCE 전략을 선택한다 하더라고요.

SEQUENCE 전략은 save할 때 Id를 먼저 파악하고 id가 포함된 insert문을 날리게 되는데, init_data.sql에 id가 포함되지 않은 데이터들을  insert해주고 있어서 문제가 발생한 것이었습니다. 그래서 id 없이도 초기 데이터를 넣어주기 위해 IDENTITY 전략을 유지하려 합니다.

pkeugine commented on src/test/java/roomescape/reservation/ReservationServiceTest.java:
```diff
@@ -0,0 +1,455 @@
+    private final MemberRepository memberRepository;
+
+    public ReservationServiceTest() {
+        reservationRepository = mock(ReservationRepository.class);
+        reservationTimeRepository = mock(ReservationTimeRepository.class);
+        themeRepository = mock(ThemeRepository.class);
+        memberRepository = mock(MemberRepository.class);
```
**Q. 기존에는 FakeRepository(FakeDao)를 구현해서 테스트에 사용했는데, join 기능을 직접 구현하는데 어려움이 있었고, Service의 비즈니스 로직을 테스트하기 위한 최소한의 stubing을 해주기 위해 Mockito를 적용했습니다.**

stub 이 가지는 장점을 몸소 체험해보셨군요 😄 
저도 stub 이 fake 에 비해 가지는 단순함, 구현 속도 증가 등의 이점을 좋게 보는 편입니다.
그렇기 때문에 제한 된 시간 안에서는 stub 을 활용하는게 더 좋다고 생각하고요.
stub, fake, mock 등에 대한 메이의 생각이 궁금하네요.

seaniiio replied:
세가지 테스트 대역에 대해 이번 미션을 구현하면서 생각을 조금 해볼 수 있었습니다. 

### stub

- 내부 구현 없이, 결과를 내가 원하는 값이 나오도록 설정합니다.
- 로직이 존재하지 않기 때문에, 특정 메서드가 특정 값을 내뱉도록 해서 원하는 기능을 테스트할 때 사용할 수 있습니다.

```java
@DisplayName("주어진 id에 해당하는 reservation 삭제한다.")
@Test
void delete1() {
    // given
    final Long id = 1L;
    given(reservationRepository.existsById(id))
            .willReturn(true);

    // when
    reservationService.deleteById(id);

    // then
    then(reservationRepository).should().deleteById(id);
}
```

- 위와 같은 경우에 유용하게 사용했습니다. 테스트하고자 하는 메서드인 `reservationService.deleteById()`가 `reservationRepository.deleteById()`를 실행하는 상황에서, `reservationRepository.deleteById()`의 결과를 조작하여 원하는 방향으로 흐름을 조정할 수 있습니다.

### fake

- 실제 프로덕션 코드와 구현 방식 자체는 달라지지만, 논리적 로직은 똑같이 흘러가도록 구현합니다. 프로덕션 코드를 테스트에 사용하지 못하는 경우, 대체적으로 사용할 수 있습니다. 그런데 작성한 내부 로직이 실제 코드와 똑같이 흘러가는지에 대한 신뢰도가 다소 떨어집니다.

### mock

- 호출이 됐는지, 몇 번 됐는지 등 확인할 때 사용합니다. 왜 사용하는지 지금은 잘 와닿지 않는데, 외부 의존성을 대체할 때 사용한다고 알고 있습니다. 외부 의존성을 mock으로 대체하고, 특정 로직에 의해 외부 기능이 호출이 잘 되는지 확인할 수 있을 것 같습니다.

---

Service는 검증해야 할 로직이 존재하는 클래스이므로, 테스트 대역을 이용한 단위 테스트가 꼭 필요하다고 느꼈습니다. stub 방식으로 테스트하니까 아주 간편하게 원하는 기능을 단위테스트 할 수 있어서 좋았어요.

pkeugine replied:
mock 은 슬라이스 테스트를 만들고, 구현이 의도대로 특정 메서드 및 클래스를 호출하는지 확인할 때 유용하죠 👍 
fake 의 경우 신뢰도 문제 때문에 테스트가 올바르게 작동하지 않거나 유지보수에 더 큰 비용이 들기도 합니다.
잘 학습해주셨습니다.

pkeugine commented on build.gradle:
```diff
@@ -21,10 +21,15 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-web'
     implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
     implementation 'org.springframework.boot:spring-boot-starter-jdbc'
+    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
```
**Q. JPA를 처음 사용해봤는데, 전환이 제대로 되었는지 궁금합니다!**

잘 만들어주셨습니다 👍 
별개로, 예전 미션에는 readme 에 todo 또는 요구사항 checklist 와 같은 문서가 포함되는 경우가 많았는데,
메이는 따로 진행사항을 관리하시지 않는지 궁금합니다.

기능 구현 및 개발도 중요하지만,
협업 할 때는 문서화가 중요하다고 생각해서요.

seaniiio replied:
점점 방탈출 프로젝트의 기능도 많아지고 복잡해지는데, 문서로 정리하도록 하겠습니다!

[docs(README.md): 기능 목록 작성](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/7cdd262c38cf093bebc585d77e8b3fd0bc1063fc)

pkeugine commented on src/test/java/roomescape/reservation/ReservationServiceTest.java:
```diff
@@ -0,0 +1,455 @@
+        private static final Optional<ReservationTime> TIME_BY_ID = Optional.of(
+                new ReservationTime(REQUEST.timeId(), LocalTime.of(12, 40)));
+        private static final Optional<Theme> THEME_BY_ID = Optional.of(
+                new Theme(REQUEST.themeId(), "야당", "야당당", "123"));
```
최근에 페어분께서 영화를 재밌게 보셨나봅니다 😂

pkeugine commented on src/main/java/roomescape/member/MemberService.java:
```diff
@@ -0,0 +1,46 @@
+
+    public List<MemberReservationResponse> readAllReservationsByMember(final LoginMember loginMember){
+        final Member member = memberRepository.findByEmail(loginMember.email())
+                .orElseThrow(() -> new MemberNotFoundException());
```
생성자도 람다 대신 아래와 같이 사용할 수 있습니다.
```suggestion
                .orElseThrow(MemberNotFoundException::new);
```

pkeugine commented on src/test/java/roomescape/reservation/ReservationServiceTest.java:
```diff
@@ -0,0 +1,455 @@
+                ReservationStatus.PENDING
+        );
+
+        @BeforeEach
+        void setUp() {
+            given(reservationTimeRepository.findById(REQUEST.timeId()))
+                    .willReturn(TIME_BY_ID);
+            given(themeRepository.findById(REQUEST.themeId()))
+                    .willReturn(THEME_BY_ID);
+            given(memberRepository.findByEmail(LOGIN_MEMBER.email()))
+                    .willReturn(MEMBER_BY_EMAIL);
+            given(reservationRepository.save(
+                    new Reservation(REQUEST.date(), MEMBER_BY_EMAIL.get(), TIME_BY_ID.get(), THEME_BY_ID.get(), ReservationStatus.PENDING)))
+                    .willReturn(RESERVATION);
+            given(reservationRepository.existsById(RESERVATION.getId()))
+                    .willReturn(true);
+            given(reservationRepository.existsByReservationTimeAndDateAndTheme(
+                    TIME_BY_ID.get(), REQUEST.date(), THEME_BY_ID.get()))
+                    .willReturn(false);
+        }
```
해당 코멘트는 개발자마다 의견이 나뉠 수 있는 내용입니다.

`@BeforeEach` 를 활용한 given 설정 좋습니다.
그런데 한 편으로는, 테스트 메서드 안에서 test fixture 를 만들고 이렇게 stub 를 설정하는 것을 선호하는 개발자도 있습니다.
아무리 중복이 있더라도 말이죠.

그 이유는, 프로젝트가 고도화 될 수록 테스트가 많아질텐데
테스트 유지보수를 하기 위해서는 그 테스트 메서드만 보고 테스트 세계관(?) 을 모두 파악할 수 있어야
유지보수하기 쉬워진다는 이유에서 입니다.
그리고 이렇게 중복이 추출되는 것부터가 테스트간 약한 의존이 형성되어 테스트 수정에 시간이 더 오래 걸릴 수 있다고 보는거죠.

이에 대한 메이의 생각은 어떤지 궁금하네요.
그리고 더 나아가 저런 입장이라면 `@BeforeEach` 를 과연 어떤 경우에 쓸까요? 🤔

seaniiio replied:
stub하는 부분은 확실히 테스트 메서드에서 처리하는 것이 맞겠다는 생각이 듭니다. 테스트 메서드만 봐서 어떤 것을 테스트하는지 명확히 확인하기 어려울 수 있다는 점은 확실히 개선해야 할 것 같아요. 또한 테스트 메서드에서 필요한 부분만 stub 처리하는게 좋을 것 같습니다.

그리고 테스트간 의존을 줄이는 방향으로 생각한다면, 상수를 선언해주는 부분을 `@BeforeEach`로 넣는게 좋을 것 같아요. 테스트간 의존을 생각한다면, 만약 어떤 테스트 메서드에서 상수를 수정하면 다른 테스트 메서드에 영향을 미칠 것 같습니다. 이런 상황을 방지하기 위해 상수를 `@BeforeEach`에서 매 테스트 메서드마다 새로 만들 것 같아요. (이 상수 마저도 테스트 메서드 내부에서 만든다면 매우 독립적인 테스트 코드를 작성할 수 있겠지만, “일반적으로 모든 테스트에서 사용 가능하도록 만든 기본적 상수”라고 본다면, 이 상수에 의해 다른 테스트가 깨지는 상황이 와닿지 않는 것 같습니다.)

seaniiio replied:
stub하는 부분은 확실히 테스트 메서드에서 처리하는 것이 맞겠다는 생각이 듭니다. 테스트 메서드만 봐서 어떤 것을 테스트하는지 명확히 확인하기 어려울 수 있다는 점은 확실히 개선해야 할 것 같아요. 또한 테스트 메서드에서 필요한 부분만 stub 처리하는게 좋을 것 같습니다.

stub을 메서드마다 해준다면, mock 객체를 만드는 작업을 `@BeforeEach`에 넣어주는 것이 좋을 것 같아요. 같은 mock 객체에 stub을 덮어쓰기보다는 매 메서드마다 새로운 mock 객체를 사용하는게 더 안전하다는 생각이 들었습니다.

그리고 테스트간 의존을 줄이는 방향으로 생각한다면, 상수를 선언해주는 부분을 `@BeforeEach`에 넣는게 좋을 것 같아요. 테스트간 의존을 생각한다면, 만약 어떤 테스트 메서드에서 상수를 수정하면 다른 테스트 메서드에 영향을 미칠 것 같습니다. 이런 상황을 방지하기 위해 상수를 `@BeforeEach`에서 매 테스트 메서드마다 새로 만들 것 같아요. (이 상수 마저도 테스트 메서드 내부에서 만든다면 매우 독립적인 테스트 코드를 작성할 수 있겠지만, “일반적으로 모든 테스트에서 사용 가능하도록 만든 기본적 상수”라고 본다면, 이 상수에 의해 다른 테스트가 깨지는 상황이 와닿지 않는 것 같습니다.)

[test: @BeforeEach에 stub fixture 설정 제거](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/fcb3c4a62f5f1156f9d079efcda0187138400c69)

pkeugine replied:
좋은 답변입니다.
`@BeforeEach` 에는 특히 테스트 격리를 위한 로직
즉, 실제 테스트를 위한 것이 아닌 테스트 환경을 조성하는 것들을 담을 수 있을 것 같네요.
예를 들어 DB 를 완전히 초기화 하는 부분이요.

pkeugine commented on src/main/java/roomescape/auth/AuthController.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final AuthService authService;
+
+    @PostMapping("/login")
```
`로그아웃 시 다시 Login 버튼이 노출됩니다.` 라는 요구사항이 저번 미션에 있었습니다.
약간 명확하지는 않은 요구사항인데요,
이번 프로젝트에 구현해보면 좋을 듯 싶습니다.

메이의 이전 프로젝트에 이미 구현해보셨다면 복습 차원에서 해보아요.

seaniiio replied:
로그아웃 기능 구현하겠습니다!

[feat: 로그아웃 기능 구현](https://github.com/woowacourse/spring-roomescape-waiting/pull/181/commits/0da90b0f973b1f96f4e9bcdd3ce4dc39e16c0abc)

pkeugine reviewed (changes requested):
안녕하세요 메이,
프로젝트 잘 구현해주셨습니다.
코멘트 한 번 확인해주세요.

바로 머지하고 다음 단계 갈 수도 있는데
해당 단계에서 배워보면 좋을 것 같은 재밌는(?) 것들이 있어서
request change 남깁니다.

seaniiio committed:
refactor: sql 쿼리 이름 기반 바인딩으로 수정

seaniiio committed:
chore: 롬복 의존성 추가

seaniiio committed:
refactor: ThemeRepository를 JpaRepository로 변경 및 FakeThemeRepository 삭제

- ThemeRepository를 Spring Data JPA 기반으로 변경하여 의존성을 단순화하고 성능 및 유지보수성을 개선.

- FakeThemeRepository 및 관련 테스트 코드 삭제.

- ThemeJdbcRepository 및 관련 초기 데이터(.sql) 삭제.

seaniiio committed:
refactor(ThemeRepository): `@Query`에서 native query 대신 JPQL 사용하도록 수정

seaniiio committed:
refactor: Member에서 Reservation으로의 일대다 단방향 연관관계 제거

seaniiio committed:
refactor(Reservation): `@AllArgsConstructor` 제거

seaniiio committed:
refactor(ReservationTime): `@AllArgsConstructor` 제거

seaniiio committed:
refactor(Theme): `@AllArgsConstructor` 제거

seaniiio committed:
refactor(Member): `@AllArgsConstructor` 제거

seaniiio committed:
refactor(ReservationTime): final 필드 제거

seaniiio committed:
refactor: `@NoArgsConstructor` 접근 제한을 protected로 설정

seaniiio committed:
docs(README.md): 기능 목록 작성

seaniiio committed:
refactor: 생성자 람다를 메서드 참조로 수정

seaniiio committed:
feat: 로그아웃 기능 구현

seaniiio committed:
test: `@BeforeEach`에 stub fixture 설정 제거

seaniiio commented:
안녕하세요 피케이!

남겨주신 리뷰 보고 많이 배우고 생각도 정리할 수 있었어요.

감사합니다 😀

pkeugine commented on README.md:
```diff
@@ -0,0 +1,42 @@
+
+- [x] 로그인할 수 있다.
+- [x] 로그아웃할 수 있다.
+- [ ] 회원가입할 수 있다.
```
문서와 애플리케이션이 같은 상태를 나타내는지,
아니라면 구현할 필요는 없는지 꼭 확인해주세요.

pkeugine reviewed (approved):
안녕하세요 메이,
리뷰나 질문이 반영하기 쉽지 않았을텐데 고생하셨습니다.
바로 다음 단계로 넘어가보아요. 💪
